homeassistant/components/apple_tv/config_flow.py

```diff
@@ -0,0 +1,403 @@
+        or credentials.get(Protocol.DMAP.value) is not None
+    )
+
+
+class AppleTVConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Apple TV."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry):
+        """Get options flow for this handler."""
+        return AppleTVOptionsFlow(config_entry)
+
+    def __init__(self):
+        """Initialize a new AppleTVConfigFlow."""
+        self.scan_result = None
+        self.atv = None
+        self.protocol = None
+        self.pairing = None
+        self.credentials = {}  # Protocol -> credentials
+
+    async def async_step_invalid_credentials(self, info):
+        """Handle initial step when updating invalid credentials."""
+        await self.async_set_unique_id(info.get(CONF_IDENTIFIER))
+
+        # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+        self.context["title_placeholders"] = {"name": info.get(CONF_NAME)}
+
+        # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+        self.context["identifier"] = self.unique_id
+        return await self.async_step_reconfigure()
+
+    async def async_step_reconfigure(self, user_input=None):
+        """Inform user that reconfiguration is about to start."""
+        if user_input is not None:
+            return await self.async_find_device_wrapper(
+                self.async_begin_pairing, allow_exist=True
+            )
+
+        return self.async_show_form(step_id="reconfigure")
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        # Be helpful to the user and look for devices
+        if self.scan_result is None:
+            self.scan_result, _ = await device_scan(None, self.hass.loop)
+
+        errors = {}
+        default_suggestion = self._prefill_identifier()
+        if user_input is not None:
+            await self.async_set_unique_id(user_input[CONF_IDENTIFIER])
+            try:
+                await self.async_find_device()
+                return await self.async_step_confirm()
+            except DeviceNotFound:
+                errors["base"] = "device_not_found"
+            except DeviceAlreadyConfigured:
+                errors["base"] = "device_already_configured"
+            except exceptions.NoServiceError:
+                errors["base"] = "no_usable_service"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+
+            # Use whatever the user entered as default value
+            default_suggestion = self.unique_id
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {vol.Required(CONF_IDENTIFIER, default=default_suggestion): str}
```

Comment:
Why do we allow the user to set this id? What does this id mean? We don't allow ip address as unique id as that's not stable. Normally the id should not be arbitrarily user configurable. It has to have a specific meaning that makes it unique and stable, eg username, gps coordinates, mac address, serial number.

Reply:
Ok, so this is legacy confusion (I only supported identifiers first). The user can enter something used to find a device. Be it an identifier, a device name or an IP address. A scan (using `pyatv`) is then performed in order to find the actual device. Unique identifiers are a core thing in `pyatv` and due to technical reasons, a device can have several unique identifiers. Any one of them can be used to find the device again and when creating the config entry, one of the found identifiers are used (might not even be the one the user entered). The IP address is *only* stored as cache and used as first attempt when connecting. If that fails, it will fall back to scanning and update the IP once a connection has been made. If IP is entered, unicast scanning is performed to find a device which is a workaround for people having issues with scanning.

https://pyatv.dev/documentation/concepts/#identifiers

This should be renamed to something else, e.g. `CONF_DEVICE` or something. I'm open for suggestions.